layer/layer_store.go

```diff
@@ -583,6 +585,10 @@ func (ls *layerStore) GetMountID(id string) (string, error) {
 func (ls *layerStore) ReleaseRWLayer(l RWLayer) ([]Metadata, error) {
 	ls.mountL.Lock()
 	m := ls.mounts[l.Name()]
+	if m != nil {
+		m.Lock()
```

Comment:
Should `RWLayer` be called one too many times(eg. refcount gets confused from error handling or live-restore) compared to the ref count then this seems like a deadlock. The first request will be stuck on `driver.Remove` second one will come in and block on `m.Lock`. After `driver.Remove` finishes it can't continue cause the second one is holding the `ls.mountL`.

Reply:
nit: on contention for the same layer (eg. last refs released together that is quite normal) this will still block the whole mountL. Better than before but not as good as what pkg/locker and similar packages do.

Reply:
I wish my built-in race detector would be as good as yours ;) Thanks!

I've replaced the ugly hack of mis-using `mountedLayer.Lock()` to per-id lock. Now we have
* `mountL sync.Mutex` which is used solely for concurrent access to `layerStore.mount map[]`
* `mountedLayer sync.Mutex` which is used solely for concurrent acccess to `mountedLayer.references map[]`
* per-id locker, to avoid name conflicts and concurrent ops on the same layer.

The whole rig seems to look more readable now (mutexes use is straightforward, no nested locks).

@tonistiigi PTALA